### PR TITLE
add index plan callback to IndexType

### PR DESCRIPTION
**Pull request by Maxxen**
This PR adds another hook to the `IndexType` class to allow indexes to control how the physical plan gets generated from a logical `CREATE INDEX` plan. 

Previously the `CreatePlan` for the `LogicalCreateIndex` operator was hard-coded to only plan `ART` indexes. Custom index types (such as those in vss and spatial) relies on optimizer extensions to "hijack" the query plan and replace the `LogicalCreateIndex` with e.g. `LogicalCreateHNSWIndex` before physical planning could begin. This hack and resulted in a lot of duplicated and very advanced code in these extensions, and also came with the unfortunate side effect that you could not create these index types at all if the optimizer was disabled.

This is just the first step in a larger extension index rework Im working on, and I want to make the interface here even tighter in the future by e.g. handling sorting/null filtering/expression type validation before we hand of control to the extension, as I think that is something that could be generalized and/or is interesting for most index types and is a bit complicated to do right now.

### CODE REVIEW

**Comment by Maxxen** [comment]
Would be great to have this on main so I can patch up spatial/vss to use this for the bug fix release.

File: src/execution/physical_plan/plan_create_index.cpp [hunk 1]
```diff
@@ -43,78 +34,11 @@ unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalCreateInde
...
-
-		// no ordering
-		physical_create_index->children.push_back(std::move(null_filter));
-	}
+	D_ASSERT(op.children.size() == 1);
+	auto table_scan = CreatePlan(*op.children[0]);
 
-	return std::move(physical_create_index);
+	PlanIndexInput input(context, op, table_scan);
+	return index_type->create_plan(input);
```
Comment (Mytherin) [inline src/execution/physical_plan/plan_create_index.cpp:41]: Can we throw an exception if `create_plan` is not set?

File: src/include/duckdb/execution/index/index_type.hpp [hunk 2]
```diff
@@ -43,14 +45,27 @@ struct CreateIndexInput {
...
+typedef unique_ptr<PhysicalOperator> (*index_plan_function_t)(PlanIndexInput &input);
+
 //! A index "type"
 class IndexType {
 public:
 	// The name of the index type
 	string name;
 
 	// Callbacks
+	index_plan_function_t create_plan;
```
Comment (Mytherin) [inline src/include/duckdb/execution/index/index_type.hpp:68]: Can we initialize these to `nullptr`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Not strictly a bug fix, but I suppose it's low risk :)

Some minor requests:

File: src/execution/physical_plan/plan_create_index.cpp (hunk 1 above)
Comment (Maxxen) [reply src/execution/physical_plan/plan_create_index.cpp:42]: I guess an assertion will do?

File: src/execution/physical_plan/plan_create_index.cpp [hunk 3]
```diff
@@ -27,6 +27,7 @@ unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalCreateInde
 	if (!index_type) {
 		throw BinderException("Unknown index type: " + op.info->index_type);
 	}
+	D_ASSERT(index_type->create_plan);
```
Comment (Mytherin) [inline src/execution/physical_plan/plan_create_index.cpp:30]: Can we make this an InternalException?

File: src/execution/physical_plan/plan_create_index.cpp (hunk 1 above)
Comment (Mytherin) [reply src/execution/physical_plan/plan_create_index.cpp:42]: I think an exception is better - calling a `nullptr` leads to a crash otherwise.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**